sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -189,6 +204,13 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
       }
   }
 
+  private def newAggOutput(aggAttribute: AttributeReference, agg: AggregateExpression) =
+    if (aggAttribute.dataType == agg.resultAttribute.dataType) {
+      aggAttribute
+    } else {
+      Cast(aggAttribute, agg.resultAttribute.dataType)
```

Comment:
I think complete and partial pushdown are different here.

For complete pushdown, we should cast to the data type of the aggregate function.
For partial pushdown, Spark will run aggregate again, so we should cast to the data type of the input of the aggregate function, so that the final data type is still the same as before.